### ISSUE TITLE
Effects of recent changes

### ISSUE THREAD

**Issue by mheyer32** [bebbo/gcc#152]
This is not a bug report per se, I am merely reporting findings and hope to confirm that they are expected in lieu of the past week's changes to gcc

The following code examples come from DoomAttack, compiled for 030 with options:
`m68k-amigaos-gcc -noixemul -msmall-code -g -ggdb -Werror -Wimplicit -Wstrict-prototypes -D__BIG_ENDIAN__ -DNORMALUNIX -DAMIGA -D__NO_NET_API -include "doomdef.h" -I./plugins/Include -I./ -DNDEBUG -Ofast -ffast-math -fomit-frame-pointer -m68030 
`
1. Sequences of  'load quick constant and add it to something loaded from memory' changed to 'load from memory, than use addq/subq to perform an operation'
```
   26b2:	moveq #1,d1
    26b4:	add.l 24 24 __stext+0x24,d1
```
changes to
```
    26a8:	move.l 24 24 __stext+0x24,d1
    26ae:	addq.l #1,d1
```

2. For some reason, the compiler prefers accessing consecutive relocs directly instead of a register. I have seen this done in multiple places. This looks like a legit regression!
```
     372:	lea e4 e4 ___main+0x1e,a0
        markpoints[i].x = -1; /* means empty*/
     378:	moveq #-1,d0
     37a:	move.l d0,(a0)+
     37c:	move.l d0,4(a0)
     380:	move.l d0,12(a0)
     384:	move.l d0,20(a0)
     388:	move.l d0,28(a0)
     38c:	move.l d0,36(a0)
     390:	move.l d0,44(a0)
     394:	move.l d0,52(a0)
     398:	move.l d0,60(a0)
     39c:	move.l d0,68(a0)
```
became
```
        markpoints[i].x = -1; /* means empty*/
     372:	moveq #-1,d0
     374:	move.l d0,e4 e4 ___main+0x1e
     37a:	move.l d0,ec ec ___main+0x26
     380:	move.l d0,f4 f4 ___main+0x2e
     386:	move.l d0,fc fc ___main+0x36
     38c:	move.l d0,104 104 ___main+0x3e
     392:	move.l d0,10c 10c ___main+0x46
     398:	move.l d0,114 114 ___main+0x4e
     39e:	move.l d0,11c 11c ___main+0x56
     3a4:	move.l d0,124 124 ___main+0x5e
     3aa:	move.l d0,12c 12c ___main+0x66
```

3. The compiler now makes use of exg!
Sequences like
```
   1d14:	movea.l a1,a0
    1d16:	movea.l d0,a1
    1d18:	move.l a0,d0
```
have become
`    1d12:	exg d0,a1`

4. The compiler tries very hard now to use dbf instead of cmp/bne for loops at the cost of more loop setup code. Is this always a win, in particular on 030?
```
     for (i = 1; i < myargc; i++)
    4d26:	move.l 24df4 24df4 _P_AimLineAttack+0x4,d0
    4d2c:	moveq #1,d1
    4d2e:	cmp.l d0,d1
    4d30:	bge.s 4d52 4d52 _D_DoomMain+0x34
    4d32:	movea.l 24df8 24df8 _P_AimLineAttack+0x8,a3
    4d38:	lea 4(a3),a0
    4d3c:	movea.l d1,a2
    4d3e:	movea.l a0,a4
    4d40:	suba.l a3,a4
        if (myargv[i][0] == '@') {
    4d42:	movea.l (a0)+,a1
    4d44:	cmpi.b #64,(a1)
    4d48:	beq.w 58c4 58c4 _D_DoomMain+0xba6
    for (i = 1; i < myargc; i++)
    4d4c:	addq.l #1,a2
    4d4e:	cmpa.l d0,a2
    4d50:	bne.s 4d3e 4d3e _D_DoomMain+0x20
```
becomes:
```
   for (i = 1; i < myargc; i++)
    4d22:	move.l 24df4 24df4 _P_SlideMove+0x884,d0
    4d28:	moveq #1,d1
    4d2a:	cmp.l d0,d1
    4d2c:	bge.s 4d5a 4d5a _D_DoomMain+0x40
    4d2e:	movea.l 24df8 24df8 _P_SlideMove+0x888,a2
    4d34:	lea 4(a2),a0
    4d38:	move.l d1,d2
    4d3a:	sub.l d0,d1
    4d3c:	move.l d1,d0
    4d3e:	not.l d0
    4d40:	movea.l a0,a3
    4d42:	suba.l a2,a3
        if (myargv[i][0] == '@') {
    4d44:	movea.l (a0)+,a1
    4d46:	cmpi.b #64,(a1)
    4d4a:	beq.w 58d0 58d0 _D_DoomMain+0xbb6
    for (i = 1; i < myargc; i++)
    4d4e:	addq.l #1,d2
    4d50:	dbf d0,4d40 4d40 _D_DoomMain+0x26
    4d54:	clr.w d0
    4d56:	subq.l #1,d0
    4d58:	bcc.s 4d40 4d40 _D_DoomMain+0x26
```

6. Byte sized values are better tracked through bftexu and move.b.  I see fewer of ` andi.l #255,d0`

**Comment by bebbo** [comment]
> 
> 
> This is not a bug report per se, I am merely reporting findings and hope to confirm that they are expected in lieu of the past week's changes to gcc
> 
> The following code examples come from DoomAttack, compiled for 030 with options:
> `m68k-amigaos-gcc -noixemul -msmall-code -g -ggdb -Werror -Wimplicit -Wstrict-prototypes -D__BIG_ENDIAN__ -DNORMALUNIX -DAMIGA -D__NO_NET_API -include "doomdef.h" -I./plugins/Include -I./ -DNDEBUG -Ofast -ffast-math -fomit-frame-pointer -m68030 `
> 
>     1. Sequences of  'load quick constant and add it to something loaded from memory' changed to 'load from memory, than use addq/subq to perform an operation'
> 
> 
> ```
>    26b2:	moveq #1,d1
>     26b4:	add.l 24 24 __stext+0x24,d1
> ```
> 
> changes to
> 
> ```
>     26a8:	move.l 24 24 __stext+0x24,d1
>     26ae:	addq.l #1,d1
> ```

this is only done for values which can be handled by addq/subq or ob 68060+ for all values. On 68080 the `addiw.l` can be used. 

>     2. For some reason, the compiler prefers accessing consecutive relocs directly instead of a register. I have seen this done in multiple places. This looks like a legit regression!
> ```
>      372:	lea e4 e4 ___main+0x1e,a0
>         markpoints[i].x = -1; /* means empty*/
>      378:	moveq #-1,d0
>      37a:	move.l d0,(a0)+
>      37c:	move.l d0,4(a0)
>      380:	move.l d0,12(a0)
>      384:	move.l d0,20(a0)
>      388:	move.l d0,28(a0)
>      38c:	move.l d0,36(a0)
>      390:	move.l d0,44(a0)
>      394:	move.l d0,52(a0)
>      398:	move.l d0,60(a0)
>      39c:	move.l d0,68(a0)
> ```
> 
> became
> 
> ```
>         markpoints[i].x = -1; /* means empty*/
>      372:	moveq #-1,d0
>      374:	move.l d0,e4 e4 ___main+0x1e
>      37a:	move.l d0,ec ec ___main+0x26
>      380:	move.l d0,f4 f4 ___main+0x2e
>      386:	move.l d0,fc fc ___main+0x36
>      38c:	move.l d0,104 104 ___main+0x3e
>      392:	move.l d0,10c 10c ___main+0x46
>      398:	move.l d0,114 114 ___main+0x4e
>      39e:	move.l d0,11c 11c ___main+0x56
>      3a4:	move.l d0,124 124 ___main+0x5e
>      3aa:	move.l d0,12c 12c ___main+0x66
> ```

this depends on free address registers. it's still done, if possible.
You can avoid it by changing the code!
 
>     3. The compiler now makes use of exg!
>        Sequences like
> 
> 
> ```
>    1d14:	movea.l a1,a0
>     1d16:	movea.l d0,a1
>     1d18:	move.l a0,d0
> ```
> 
> have become
> ` 1d12: exg d0,a1`

nice, eh?

>     4. The compiler tries very hard now to use dbf instead of cmp/bne for loops at the cost of more loop setup code. Is this always a win, in particular on 030?

it's almost always a win, consider the loop end with 2 or 3 insns vs. only 1 insn.
It's a win on all from 68000 to 68080

> ```
>      for (i = 1; i < myargc; i++)
>     4d26:	move.l 24df4 24df4 _P_AimLineAttack+0x4,d0
>     4d2c:	moveq #1,d1
>     4d2e:	cmp.l d0,d1
>     4d30:	bge.s 4d52 4d52 _D_DoomMain+0x34
>     4d32:	movea.l 24df8 24df8 _P_AimLineAttack+0x8,a3
>     4d38:	lea 4(a3),a0
>     4d3c:	movea.l d1,a2
>     4d3e:	movea.l a0,a4
>     4d40:	suba.l a3,a4
>         if (myargv[i][0] == '@') {
>     4d42:	movea.l (a0)+,a1
>     4d44:	cmpi.b #64,(a1)
>     4d48:	beq.w 58c4 58c4 _D_DoomMain+0xba6
>     for (i = 1; i < myargc; i++)
>     4d4c:	addq.l #1,a2
>     4d4e:	cmpa.l d0,a2
>     4d50:	bne.s 4d3e 4d3e _D_DoomMain+0x20
> ```
> 
> becomes:
> 
> ```
>    for (i = 1; i < myargc; i++)
>     4d22:	move.l 24df4 24df4 _P_SlideMove+0x884,d0
>     4d28:	moveq #1,d1
>     4d2a:	cmp.l d0,d1
>     4d2c:	bge.s 4d5a 4d5a _D_DoomMain+0x40
>     4d2e:	movea.l 24df8 24df8 _P_SlideMove+0x888,a2
>     4d34:	lea 4(a2),a0
>     4d38:	move.l d1,d2
>     4d3a:	sub.l d0,d1
>     4d3c:	move.l d1,d0
>     4d3e:	not.l d0
>     4d40:	movea.l a0,a3
>     4d42:	suba.l a2,a3
>         if (myargv[i][0] == '@') {
>     4d44:	movea.l (a0)+,a1
>     4d46:	cmpi.b #64,(a1)
>     4d4a:	beq.w 58d0 58d0 _D_DoomMain+0xbb6
>     for (i = 1; i < myargc; i++)
>     4d4e:	addq.l #1,d2
>     4d50:	dbf d0,4d40 4d40 _D_DoomMain+0x26
>     4d54:	clr.w d0
>     4d56:	subq.l #1,d0
>     4d58:	bcc.s 4d40 4d40 _D_DoomMain+0x26
> ```
> 
>     5. Byte sized values are better tracked through bftexu and move.b.  I see fewer of ` andi.l #255,d0`

the optimizers tries its best^^

**Comment by mheyer32** [comment]
> this depends on free address registers. it's still done, if possible.
> You can avoid it by changing the code!

Can you elaborate?

**Comment by bebbo** [comment]
please revisit the generated code again

**Comment by mheyer32** [comment]
Confirmed. The base-register-relative addressing of symbols is back!

One thing that caught my eye:

```
    memcpy(cmd, base, sizeof(*cmd));
    bc5a:	movea.l a2,a0
    bc5c:	movea.l d0,a1
    bc5e:	move.b (a1)+,(a0)+
    bc60:	move.b (a1)+,(a0)+
    bc62:	move.b (a1)+,(a0)+
    bc64:	move.b (a1)+,(a0)+
    bc66:	move.b (a1)+,(a0)+
    bc68:	move.b (a1)+,(a0)+
    bc6a:	move.b (a1)+,(a0)+
    bc6c:	move.b (a1),(a0)
```
became:

```
    bc5e:	movea.l d0,a0
    bc60:	move.b (a0)+,(a2)
    bc62:	move.b (a0)+,1(a2)
    bc66:	move.b (a0)+,2(a2)
    bc6a:	move.b (a0)+,3(a2)
    bc6e:	move.b (a0)+,4(a2)
    bc72:	move.b (a0)+,5(a2)
    bc76:	move.b (a0)+,6(a2)
    bc7a:	move.b (a0),7(a2)
    memcpy(cmd, base, sizeof(*cmd));
```
Instruction size for the latter is larger. I did not check which one is faster, but my hunch would be the original?

**Comment by bebbo** [comment]
so basically this: http://franke.ms/cex/z/1Gzeqn

**Comment by mheyer32** [comment]
Yes. Interestingly, gcc 2.95 generates even better code, by doing 2 32bit moves. Can haz in gcc 6.5 plz? :-D

**Comment by bebbo** [comment]
Good luck with odd addresses then^^

**Comment by mheyer32** [comment]
Depends on the natural alignment/size of the copied struct - things the compiler can figure out once it emits code for memcpy. 
Adding '-fpack-struct' option forces gcc 2.95 to use the byte sized copies, too.

On top, for anything above 68000 odd addresses are not even an issue, are they?

**Comment by bebbo** [comment]
> 
> 
> Depends on the natural alignment/size of the copied struct - things the compiler can figure out once it emits code for memcpy.
> Adding '-fpack-struct' option forces gcc 2.95 to use the byte sized copies, too.
> 
> On top, for anything above 68000 odd addresses are not even an issue, are they?

Only 68000 is affected. But I don't know what happens if the address refers to hardware registers...

**Comment by bebbo** [comment]
The challenge is to keep this too: http://franke.ms/cex/z/8cPGK7

**Comment by bebbo** [comment]
with `-m68020` you get what you can haz^^ http://franke.ms/cex/z/M88jKb

**Comment by mheyer32** [comment]
Awesome!
```
   base = I_BaseTiccmd(); /* empty, or external driver*/
    bc54:	jsr 13246 13246 _I_BaseTiccmd
    memcpy(cmd, base, sizeof(*cmd));
    bc5a:	movea.l a2,a0
    bc5c:	movea.l d0,a1
    bc5e:	move.b (a1)+,(a0)+
    bc60:	move.b (a1)+,(a0)+
    bc62:	move.b (a1)+,(a0)+
    bc64:	move.b (a1)+,(a0)+
    bc66:	move.b (a1)+,(a0)+
    bc68:	move.b (a1)+,(a0)+
    bc6a:	move.b (a1)+,(a0)+
    bc6c:	move.b (a1),(a0)
```
turned into
```
    base = I_BaseTiccmd(); /* empty, or external driver*/
    bc48:	jsr 1322e 1322e _I_BaseTiccmd
    memcpy(cmd, base, sizeof(*cmd));
    bc4e:	movea.l a2,a0
    bc50:	movea.l d0,a1
    bc52:	move.l (a1)+,(a0)+
    bc54:	move.l (a1),(a0)
```

**Comment by mheyer32** [comment]
The compiler seems to be very conservative, though, when to replace a call to memcpy() with an inline copy (and above optimization). 
What limitations does memcpy have, assuming the size is relatively small and known at compile time?

**Comment by bebbo** [comment]
> 
> 
> The compiler seems to be very conservative, though, when to replace a call to memcpy() with an inline copy (and above optimization).
> What limitations does memcpy have, assuming the size is relatively small and known at compile time?

gcc counts the number of moves - before combining these (b->w->l). The limits are currently 15 moves and 6 moves with `-Os`.

**Comment by mheyer32** [comment]
Now that the compiler can move up to 4 times the data with the same number of instructions, does it it make sense to increase these limits by 4x?

**Comment by bebbo** [comment]
> 
> 
> Now that the compiler can move up to 4 times the data with the same number of instructions, does it it make sense to increase these limits by 4x?

feel free to patch you local version and start experiments

**Comment by mheyer32** [comment]
ok, where would I patch that?

**Comment by bebbo** [comment]
> 
> 
> ok, where would I patch that?

/gcc/gcc/config/m68k/m68kamigaos.h:
```
#undef MOVE_RATIO
#undef CLEAR_RATIO
//#define MOVE_RATIO(speed) ((speed) ? 15 : 3)
//#define CLEAR_RATIO(speed) ((speed) ? 15 :3)
#define MOVE_RATIO(speed) ((speed) ? 15 : 6)
#define CLEAR_RATIO(speed) ((speed) ? 15 : 6)
```

**Comment by bebbo** [comment]
> I did play around with it a bit but I don't understand the ratio yet. Is it about the number of bytes moved or the number of instructions taken?

AFAIK it's the size in bytes. The first value is for -O1, -O2, -O3 and -Ofast the latter for -Os, the 2nd should stay low since size matters more than speed.
 
> One thing this experiment revealed is that the move combines are not working 100% yet for stuff on the stack:
> 
> ```
>     memcpy(save_p, name2, VERSIONSIZE);
>     cf50:	pea 10 10 __stext+0x10
>     cf54:	move.l a3,-(sp)
>     cf56:	move.l (a6),-(sp)
>     cf58:	jsr 44c0c 44c0c _memcpy
> ```
> 
> was turned into
> 
> ```
>  memcpy(save_p, name2, VERSIONSIZE);
>     d386:	movea.l d4,a1
>     d388:	movea.l 26034 26034 _P_BoxOnLineSide+0x1dc,a0
>     d38e:	move.b (a1),(a0)+
>     d390:	move.b 65(sp),(a0)+
>     d394:	move.b 66(sp),(a0)+
>     d398:	move.b 67(sp),(a0)+
>     d39c:	move.b 68(sp),(a0)+
>     d3a0:	move.b 69(sp),(a0)+
>     d3a4:	move.b 70(sp),(a0)+
>     d3a8:	move.b 71(sp),(a0)+
>     d3ac:	move.b 72(sp),(a0)+
>     d3b0:	move.b 73(sp),(a0)+
>     d3b4:	move.b 74(sp),(a0)+
>     d3b8:	move.b 75(sp),(a0)+
>     d3bc:	move.b 76(sp),(a0)+
>     d3c0:	move.b 77(sp),(a0)+
>     d3c4:	move.b 78(sp),(a0)+
>     d3c8:	move.b 79(sp),(a0)
> ```
> 
> when I increased the ratio to 64, 24

well, every variant of instructions needs a special treatment...

**Comment by mheyer32** [comment]
> well, every variant of instructions needs a special treatment...

Yep, no rush. Take care of it if and when you feel like it.  I can see how this would lead into a rabbit hole.

Similar issue here:

```
   13002:	lea -37(a5),a3
                memcpy(id, C2P->id, 4);
   13006:	move.b 8(a2),(a3)
   1300a:	move.b 9(a2),-36(a5)
   13010:	move.b 10(a2),-35(a5)
   13016:	move.b 11(a2),-34(a5)
```

**Comment by mheyer32** [comment]
digging some more grepping shows:

tree-inline.c

```
  if (size < 0 || size > MOVE_MAX_PIECES * MOVE_RATIO (speed_p))
    /* Cost of a memcpy call, 3 arguments and the call.  */
    return 4;
  else
    return ((size + MOVE_MAX_PIECES - 1) / MOVE_MAX_PIECES);
```

where MOVE_MAX_PIECES boils down to MOVE_MAX which is defined as 4 on m68k

So I guess the RATIO refers to longwords already?

This code seems to woefully underestimate the cost of memcpy on the Amiga. A memcpy call there looks like this:

```
00044c0c 00044c0c _memcpy:
   44c0c:	move.l a6,-(sp)
   44c0e:	move.l d2,-(sp)
   44c10:	move.l 12(sp),d2
   44c14:	movea.l 1334 1334 _AM_Responder+0x608,a6  //SysBase
   44c1a:	movea.l 16(sp),a0
   44c1e:	move.l 20(sp),d0
   44c22:	movea.l d2,a1
   44c24:	jsr -624(a6)
   44c28:	move.l d2,d0
   44c2a:	move.l (sp)+,d2
   44c2c:	movea.l (sp)+,a6
   44c2e:	rts
```

(ignoring all the setup and save/restore that CopyMem will still need to do)

**Comment by bebbo** [comment]
MOVE_RATIO is no longer used :-) please have a look

**Comment by mheyer32** [comment]
Sweet!

I see that not only memcpy has been optimized to be inlined whenever possible, but also memset had been optimized the same way :-D

But something is broken, DoomAttack wouldn't run anymore:
![image](https://user-images.githubusercontent.com/33539855/112786725-c2130b00-900b-11eb-89d0-92c3e8ab646c.png)

**Comment by mheyer32** [comment]
Update: it is the memset() that is somehow broken

**Comment by bebbo** [comment]
uhm - here it runs - if you are using devel1: I just pushed a cleanup fix...

**Comment by mheyer32** [comment]
I tested  gcc-6-branch @49524931825470686e60da0ef33de85edbae4d25

**Comment by bebbo** [comment]
please attach your binary and file the options you are using

**Comment by bebbo** [comment]
and test 2c29a22e286b2fabb66620a27721f84f78cd078e

**Comment by bebbo** [comment]
ok - I can reproduce it

**Comment by mheyer32** [comment]
Ha, you're faster in testing than me responding X-D here

**Comment by bebbo** [comment]
`move.l #$fefefeff,(a0)+`

... ouch

**Comment by bebbo** [comment]
fix queued: 5ef70c592b77834687edf3b5f915328eacfa4701

**Comment by bebbo** [comment]
fix is live: 91f455cfb7238a3f88a92150dad3ce530ba9bd1d

**Comment by mheyer32** [comment]
Thanks, that fixed it!
I'm quite happy with the memcpy/memset code. Looks beautiful now :-)
